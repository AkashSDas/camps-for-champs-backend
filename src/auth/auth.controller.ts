import { Request, Response } from "express";

// eslint-disable-next-line prettier/prettier
import { Body, Controller, Get, Post, Req, Res, UseGuards } from "@nestjs/common";

import { AuthService } from "./auth.service";
import { SignupDto } from "./dto";
import { LoginDto } from "./dto/login.dto";
import { AccessTokenGuard, RefreshTokenGuard } from "./guard";

@Controller("/v1/auth")
export class AuthController {
  constructor(private service: AuthService) {}

  // ================================
  // SIGNUP
  // ================================

  @Post("/signup")
  async signup(
    @Body() dto: SignupDto,
    @Res({ passthrough: true }) res: Response,
  ) {
    return await this.service.signup(dto, res);
  }

  // ================================
  // LOGIN
  // ================================

  @Post("/login")
  async login(
    @Body() dto: LoginDto,
    @Res({ passthrough: true }) res: Response,
  ) {
    return await this.service.login(dto, res);
  }

  @Get("/access-token")
  @UseGuards(RefreshTokenGuard)
  accessToken(@Req() req: Request) {
    return this.service.accessToken(req);
  }

  // ================================
  // OTHER
  // ================================

  @Get("/logout")
  logout(@Req() req: Request, @Res({ passthrough: true }) res: Response) {
    return this.service.logout(req, res);
  }

  // ================================
  // TEST
  // ================================

  @Get("/test")
  @UseGuards(AccessTokenGuard)
  test(@Req() req: Request) {
    return { user: req.user, message: "üåç Secret operation" };
  }
}
