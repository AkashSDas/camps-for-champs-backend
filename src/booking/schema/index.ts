export * from "./member.schema";
export * from "./booking.schema";
