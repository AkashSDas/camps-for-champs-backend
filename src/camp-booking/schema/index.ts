export * from "./camp-booking.schema";
export * from "./member.schema";
