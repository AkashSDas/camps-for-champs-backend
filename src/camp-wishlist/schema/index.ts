export * from "./camp-wishlist.schema";
