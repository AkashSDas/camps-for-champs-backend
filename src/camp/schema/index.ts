export * from "./camp.schema";
export * from "./location.schema";
export * from "./image.schema";
export * from "./cancellation-policy.schema";
