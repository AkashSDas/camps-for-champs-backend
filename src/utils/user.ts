export enum UserRole {
  BASE = "base",
  ADMIN = "admin",
  ORGANIZATION = "organization",
}
