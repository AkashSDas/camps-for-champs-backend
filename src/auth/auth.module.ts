import { AuthController } from "./auth.controller";
import { AuthService } from "./auth.service";
import { JwtModule } from "@nestjs/jwt";
import { Module } from "@nestjs/common";
import { UserModule } from "../user/user.module";
import {
  AccessTokenStrategy,
  GoogleLoginStrategy,
  GoogleSignupStrategy,
  RefreshTokenStrategy,
} from "./strategy";

@Module({
  imports: [UserModule, JwtModule.register({})],
  controllers: [AuthController],
  providers: [
    AuthService,
    AccessTokenStrategy,
    RefreshTokenStrategy,
    GoogleSignupStrategy,
    GoogleLoginStrategy,
  ],
})
export class AuthModule {}
