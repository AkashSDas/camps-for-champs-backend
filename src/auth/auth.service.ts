import { Types } from "mongoose";

// eslint-disable-next-line prettier/prettier
import { Injectable } from "@nestjs/common";
import { JwtService } from "@nestjs/jwt";

import { User } from "../user/schema/index";
import { UserRepository } from "../user/user.repository";
import { EmailAndPasswordLoginDto, EmailAndPasswordSignupDto } from "./dto";

type EmailAndPasswordLogin = {
  user?: User & { _id: Types.ObjectId };
  error?: Error;
  accessToken?: string;
  refreshToken?: string;
};

@Injectable()
export class AuthService {
  constructor(private repository: UserRepository, private jwt: JwtService) {}

  // =====================================
  // Signup
  // =====================================

  async emailAndPasswordSignup(dto: EmailAndPasswordSignupDto) {
    // Create user
    // Using arrow function to use AuthSerivce's this context
    var user: (User & { _id: Types.ObjectId }) | Error = await (async () => {
      try {
        let user = await this.repository.create({
          email: dto.email,
          passwordDigest: dto.password,
        });

        if (!user) return Error("Failed to create user");
        return user;
      } catch (error) {
        if (error instanceof Error) {
          let msg = error.message;
          if (msg == "Email already exists" || msg == "Duplicate fields") {
            return Error("User already exists");
          }
        }

        return Error("Failed to create user");
      }
    })();

    // Send result
    if (user instanceof Error) {
      return { user: null, error: user, accessToken: null, refreshToken: null };
    } else {
      let accessToken = user.getAccessToken(this.jwt);
      let refreshToken = user.getRefreshToken(this.jwt);
      user.passwordDigest = undefined; // rm password has from response
      return { user, accessToken, refreshToken, error: null };
    }
  }

  // =====================================
  // Login
  // =====================================

  async emailAndPasswordLogin(
    dto: EmailAndPasswordLoginDto,
  ): Promise<EmailAndPasswordLogin> {
    var user = await this.repository.getWithSelect(
      { email: dto.email },
      "+passwordDigest",
    );

    if (!user) return { error: Error("User not found") };

    {
      // Verify password
      let isPasswordValid = await user.verifyPassword(dto.password);
      user.passwordDigest = undefined; // rm password digest from response
      if (!isPasswordValid) return { user, error: Error("Wrong password") };
    }

    {
      // Login user
      let accessToken = user.getAccessToken(this.jwt);
      let refreshToken = user.getAccessToken(this.jwt);
      return { accessToken, refreshToken, user };
    }
  }
}
