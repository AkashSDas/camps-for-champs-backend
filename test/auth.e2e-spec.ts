import * as cookieParser from "cookie-parser";
import * as session from "express-session";
import { MongoMemoryServer } from "mongodb-memory-server";
import { connect, Connection, Document, Model } from "mongoose";
import * as request from "supertest";

import { INestApplication, ValidationPipe } from "@nestjs/common";
import { ConfigModule } from "@nestjs/config";
import { JwtService } from "@nestjs/jwt";
// eslint-disable-next-line prettier/prettier
import { getModelToken, MongooseModule } from "@nestjs/mongoose";
import { Test, TestingModule } from "@nestjs/testing";

import { AuthModule } from "../src/auth/auth.module";
import { User } from "../src/user/schema";
import { UserModule } from "../src/user/user.module";

describe("auth", () => {
  let app: INestApplication;
  let mongod: MongoMemoryServer;
  let mongoConnection: Connection;
  let userModel: Model<User & Document>;
  let jwt: JwtService;

  beforeAll(async () => {
    mongod = await MongoMemoryServer.create();
    const uri = mongod.getUri();
    mongoConnection = (await connect(uri)).connection;

    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({ isGlobal: true }),
        MongooseModule.forRoot(uri),
        AuthModule,
        UserModule,
      ],
      providers: [{ provide: getModelToken(User.name), useValue: userModel }],
    }).compile();

    userModel = module.get<Model<User & Document>>(getModelToken(User.name));
    jwt = module.get<JwtService>(JwtService);

    app = module.createNestApplication();
    app.useGlobalPipes(new ValidationPipe({ whitelist: true }));
    app.use(cookieParser());
    app.use(
      session({
        secret: process.env.COOKIE_SESSION_SECRET,
        resave: true,
        saveUninitialized: true,
      }),
    );

    await app.init();
  });

  afterAll(async () => {
    await mongoConnection.dropDatabase();
    await mongoConnection.close();
    await mongod.stop();
  });

  afterEach(async () => {
    await userModel.deleteMany({});
  });

  // =====================================
  // Signup
  // =====================================

  describe("email-signup", () => {
    describe("when user doesn't exists", () => {
      it("then successfully do a signup", async () => {
        let response = await request(app.getHttpServer())
          .post("/v2/auth/email-signup")
          .send({ email: "james@gmail.com", password: "testingTEST@123" });

        expect(response.status).toEqual(201);
        expect(response.body).toMatchObject({
          user: { email: "james@gmail.com" },
        });

        expect(response.body).toMatchSnapshot({
          accessToken: expect.any(String),
          user: {
            ...response.body.user,
            _id: expect.any(String),
            userId: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          },
        });

        expect(
          response.headers["set-cookie"].find((cookie: string) =>
            cookie.includes("refreshToken"),
          ),
        ).toBeDefined();
      });
    });

    describe("when user already exists", () => {
      beforeEach(async () => {
        await userModel.create({
          email: "james@gmail.com",
          passwordDigest: "testingTEST@123",
        });
      });

      it("then given an 400 error", async () => {
        let response = await request(app.getHttpServer())
          .post("/v2/auth/email-signup")
          .send({ email: "james@gmail.com", password: "testingTEST@123" });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

  // =====================================
  // Login
  // =====================================

  describe("email-login", () => {
    describe("when user doesn't exists", () => {
      it("then give a 404 response", async () => {
        let response = await request(app.getHttpServer())
          .post("/v2/auth/email-login")
          .send({ email: "james@gmail.com", password: "testingTEST@123" });

        expect(response.status).toEqual(404);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe("when user exists", () => {
      beforeEach(async () => {
        await userModel.create({
          email: "james@gmail.com",
          passwordDigest: "testingTEST@123",
        });
      });

      it("then successfully login", async () => {
        let response = await request(app.getHttpServer())
          .post("/v2/auth/email-login")
          .send({ email: "james@gmail.com", password: "testingTEST@123" });

        expect(response.status).toEqual(200);

        expect(response.body).toMatchSnapshot({
          accessToken: expect.any(String),
          user: {
            ...response.body.user,
            _id: expect.any(String),
            userId: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          },
        });

        expect(
          response.headers["set-cookie"].find((cookie: string) =>
            cookie.includes("refreshToken"),
          ),
        ).toBeDefined();
      });
    });
  });

  describe("access-token", () => {
    let refreshToken: string;

    beforeEach(async () => {
      let user = await userModel.create({
        email: "james@gmail.com",
        passwordDigest: "testingTEST@123",
      });

      refreshToken = user.getRefreshToken(jwt);
    });

    describe("when refresh token is valid", () => {
      it("then successfully get a new access token", async () => {
        let response = await request(app.getHttpServer())
          .get("/v2/auth/access-token")
          .set("Cookie", [`refreshToken=${refreshToken}`]);

        expect(response.status).toEqual(200);

        expect(response.body).toMatchSnapshot({
          accessToken: expect.any(String),
          user: {
            ...response.body.user,
            _id: expect.any(String),
            userId: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          },
        });
      });
    });
  });
});
