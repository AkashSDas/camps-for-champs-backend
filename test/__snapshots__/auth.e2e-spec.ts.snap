// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/v2/auth email signup successful signup with valid data 1`] = `
Object {
  "accessToken": Any<String>,
  "user": Object {
    "__v": 0,
    "_id": Any<String>,
    "active": false,
    "banned": false,
    "createdAt": Any<String>,
    "email": "james@gmail.com",
    "oauthProviders": Array [],
    "roles": Array [
      "base",
    ],
    "updatedAt": Any<String>,
    "userId": Any<String>,
    "verified": false,
  },
}
`;

exports[`/v2/auth email signup user already exists if user already then given an user exists error 1`] = `
Object {
  "error": "Bad Request",
  "message": "User already exists",
  "statusCode": 400,
}
`;
